base/system/winlogon/sas.c

```diff
@@ -279,153 +279,305 @@ PlaySoundRoutine(
+    if (lRegStatus != ERROR_SUCCESS)
+    {
+        return 0;
+    }
+
+    /* Open .Current */
+    lRegStatus = RegOpenKeyExW(hRegKey, L".Current", 0, KEY_QUERY_VALUE, &hRegSnd);
+
+    if (lRegStatus != ERROR_SUCCESS)
+    {
+        /* If fail then open .Default */
+        lRegStatus = RegOpenKeyExW(hRegKey, L".Default", 0, KEY_QUERY_VALUE, &hRegSnd);
+
+        RegCloseKey(hRegKey);
+
+        if (lRegStatus != ERROR_SUCCESS)
         {
-            TRACE("QueryServiceStatusEx failed (%x)\n", GetLastError());
-            break;
+            return 0;
         }
+    }
 
-        if (Info.dwCurrentState == SERVICE_RUNNING)
-            break;
+    RegCloseKey(hRegKey);
 
-        Sleep(1000);
+    /* Get registry value size */
+    lRegStatus = RegQueryValueExW(hRegSnd, NULL, NULL, &dwRegValType, NULL, &dwRegValSize);
 
-    } while (Index++ < 20);
+    /* Check whether the type is valid */
+    if (lRegStatus != ERROR_SUCCESS || (dwRegValType != REG_SZ && dwRegValType != REG_EXPAND_SZ))
+    {
+        RegCloseKey(hRegSnd);
+        return 0;
+    }
 
-    CloseServiceHandle(hService);
-    CloseServiceHandle(hSCManager);
+    /* Allocate buffer for registry value */
+    lpRegVal = HeapAlloc(GetProcessHeap(), 0, (SIZE_T)dwRegValSize);
 
-    /* If wdmaud is not running exit */
-    if (Info.dwCurrentState != SERVICE_RUNNING)
+    if (!lpRegVal)
     {
-        WARN("wdmaud has not started!\n");
+        RegCloseKey(hRegSnd);
         return 0;
     }
 
-    /* Sound subsystem is running. Play logon sound. */
-    TRACE("Playing logon sound: %ls\n", wszDest);
-    PlaySoundRoutine(wszDest, TRUE, SND_FILENAME);
+    /* Get registry value */
+    RegQueryValueExW(hRegSnd, NULL, NULL, &dwRegValType, (LPBYTE)lpRegVal, &dwRegValSize);
+
+    RegCloseKey(hRegSnd);
+
+    /* Get full sound path length (including the terminating null character) */
+    dwSndPathLen = ExpandEnvironmentStringsW(lpRegVal, NULL, 0);
+
+    if (dwSndPathLen == 0)
+    {
+        HeapFree(GetProcessHeap(), 0, lpRegVal);
+        return 0;
+    }
+
+    /* Allocate buffer for sound path */
+    lpSndPath = HeapAlloc(GetProcessHeap(), 0, (SIZE_T)(dwSndPathLen * sizeof(WCHAR)));
+
+    if (!lpSndPath)
+    {
+        HeapFree(GetProcessHeap(), 0, lpRegVal);
+        return 0;
+    }
+
+    ExpandEnvironmentStringsW(lpRegVal, lpSndPath, dwSndPathLen);
+
+    TRACE("Playing logoff sound: %ls\n", lpSndPath);
+
+    /* Play sound */
+    PlaySoundRoutine(lpSndPath, TRUE, SND_FILENAME | SND_NODEFAULT);
+
+    HeapFree(GetProcessHeap(), 0, lpRegVal);
+    HeapFree(GetProcessHeap(), 0, lpSndPath);
+
     return 0;
 }
 
-static
-VOID
-PlayLogonSound(
-    IN OUT PWLSESSION Session)
+static void PlayLogonSound(PWLSESSION Session)
+{
+    HANDLE hThread;
+
+    hThread = CreateThread(NULL, 0, (LPTHREAD_START_ROUTINE)PlayLogonSoundThread, (LPVOID)Session, 0, NULL);
+
+    if (hThread)
+    {
+        CloseHandle(hThread);
+    }
+}
+
+static void PlayLogoffSound(PWLSESSION Session)
```

Comment:
There seems to be quite redundancy between these functions, whose behaviour just differ by only **one** thing (the thread function)!

Reply:
Updated.